Content.Server/Database/ServerDbBase.cs

```diff
@@ -1720,6 +1720,58 @@ public async Task<bool> RemoveJobWhitelist(Guid player, ProtoId<JobPrototype> jo
 
         #endregion
 
+        # region IPIntel
+
+        public async Task<bool> UpsertIPIntelCache(DateTime time, IPAddress ip, float score)
+        {
+            await using var db = await GetDb();
+
+            var existing = await db.DbContext.IPIntelCache
+                .Where(w => ip.Equals(w.Address))
+                .SingleOrDefaultAsync();
+
+            if (existing == null)
+            {
+                var newCache = new IPIntelCache
+                {
+                    Time = time,
+                    Address = ip,
+                    Score = score,
+                };
+                db.DbContext.IPIntelCache.Add(newCache);
+            }
+            else
+            {
+                existing.Time = time;
+                existing.Score = score;
+            }
+
+            await db.DbContext.SaveChangesAsync();
+            return true;
+        }
```

Comment:
Can this cause database conflicts if two IP addresses are attempted to be saved at once? I'm pretty sure you need logic to retry since select + insert is not an atomic operation.